### PR TITLE
wrappers: generate D-Bus service activation files

### PR DESCRIPTION
**Pull request by jhenstridge**
This is the last part split off from PR #6258.  At present, it builds off the tip of PR #8861, so it probably shouldn't be reviewed until that is merged.  It does the following:

1. Generate service activation files for each activatable D-Bus name in the snap.  This is done through the `wrappers` package now rather than as part of interface handling so that they can be written after the systemd service files.

2. If `bus-name` is empty on a `daemon: dbus` app, use the bus name from the last `activates-on` slot.  In the original PR, `bus-name` was removed entirely, but `daemon: dbus` is useful for non-activatable apps like NetworkManager.

3. Bring over the test snaps and spread tests from the original PR.

The original conflict detection code is gone in favour of PR #8860, but we still read the `X-Snap` key from service files to ensure we clean up strays on install/removal to ensure we don't get confused by modifications to `snap try` snaps.  As we are only using this to find the snap's own service activation files, it shouldn't represent the same concurrency problem the old code did.

### CODE REVIEW

**Comment by zyga** [comment]
It seems tests need some adjustments:
```
distro_install_package dbus-user-session
quiet: zypper install -y --allow-downgrade --force-resolution dbus-user-session
quiet: exit status 104. Output follows:
Loading repository data...
Reading installed packages...
'dbus-user-session' not found in package names. Trying capabilities.
No provider of 'dbus-user-session' found.
quiet: end of output.
```

**Comment by jhenstridge** [comment]
Yep.  This is not yet ready for review (in part because it is built on top of a not yet merged PR).  I pushed it up as a draft to get some feedback from CI.

**Comment by jhenstridge** [comment]
One of the causes of the current test failures is that the repacked core snap is missing the additional D-Bus config.  This will resolve itself once #8861 is merged and a new core snap is published to edge, so I'm not sure whether it is worth diving in to `update_core_snap_for_classic_reexec`.

**Comment by jhenstridge** [comment]
@jdstrand: I think this probably needs a security review.  I know you had done a similar review of #6258, but the implementation has changed somewhat while splitting the feature.  There's less to review this time around though, as some of the base features (which don't change the behaviour of snaps) have been merged.

File: wrappers/services_gen_test.go
Comment (pedronis) [inline wrappers/services_gen_test.go:270]: should we keep a test where only bus-name is specified?

File: wrappers/dbus.go
Comment (pedronis) [inline wrappers/dbus.go:164]: this will not put in place the previous state of session service files, it will remove old and new, I suppose it's what you want/expect?  but is not very clear

File: tests/main/dbus-activation-name-conflict/task.yaml
Comment (pedronis) [inline tests/main/dbus-activation-name-conflict/task.yaml:25]: should this check that the error is about what we expect and not something else unrelated?

File: overlord/snapstate/backend/link_test.go
Comment (jhenstridge) [reply overlord/snapstate/backend/link_test.go:138]: There's no requirement that an activatable service use `daemon: dbus`.  It is usually a good idea, since it provides notification that the daemon has started without the extra privileges `daemon: notify` requires.

On the `dbus-daemon` side, it will issue the method call as soon as the bus name has been acquired, which could be before or after systemd thinks the daemon has started.

File: wrappers/dbus.go
Comment (jhenstridge) [reply wrappers/dbus.go:164]: I've changed the comment to make it clear it is removing files in the error case.

File: wrappers/services_gen_test.go
Comment (jhenstridge) [reply wrappers/services_gen_test.go:270]: Good idea.  I've added an extra test case.

File: tests/main/dbus-activation-name-conflict/task.yaml
Comment (jhenstridge) [reply tests/main/dbus-activation-name-conflict/task.yaml]: I've updated the test to check that the error message is about the name conflict.

**Comment by jhenstridge** [comment]
The failures on various Ubuntu releases seem to be due to weirdness in the edge channel of core not matching up with master at the moment.  When that is resolved, they should start passing again.

**Comment by alexmurray** [comment]
Thanks @jhenstridge this looks good to me from a security point-of-view. @jdstrand do you still want to take a look at this as well?

File: wrappers/dbus_test.go
Comment (jdstrand) [inline wrappers/dbus_test.go:113]: @alexmurray asked if I wanted to look at this any more. I didn't look at this very closely and trust @alexmurray's review, but one thing did come to mind: I don't see in the PR negative tests for crafted values for these fields in the activation file. I think we should be ok since there are checks elsewhere that for example, make sure that the dbus name is valid, but having a check or two with invalid values for each field we are writing out would be a good idea to ensure all the dots stay connected going forward (the exhaustive checks can of course remain in the parsing code).

**Comment by jdstrand** [comment]
I'm going to remove 'Needs Security Review' since @alexmurray gave his review, but please consider my last comment.

File: wrappers/dbus_test.go
Comment (jhenstridge) [reply wrappers/dbus_test.go:113]: For what it is worth, the data we're substituting in here has all been validated as part of loading the `snap.yaml`.

If the bus name in the slot definition does not pass `ValidateDBusBusName()`, then the dbus interface's `BeforePrepareSlot` method will fail.  This causes the slot to be marked as bad and deleted from the snap.  That in turn means we will not generate an activation file wrapper for the slot.

Comment (jhenstridge) [reply wrappers/dbus_test.go:114]: I've added a unit test to demonstrate how invalid dbus slots will be ignored.

File: wrappers/dbus.go
Comment (stolowski) [inline wrappers/dbus.go:35]:
```suggestion
	// The D-Bus service activation file format is defined as part
```

File: wrappers/dbus.go
Comment (stolowski) [inline wrappers/dbus.go:173]: Thanks for this comment, I was naively thinking this could be optimized by checking if we have any .ActivatesOn slots but yes, not way around it. I wonder if we could have a trivial snap try spread test, but maybe it's a material for a followup at this point.

File: wrappers/dbus_test.go
Comment (stolowski) [inline wrappers/dbus_test.go:154]: Nice

**Review by stolowski — APPROVED** [review]
Looks very good, thanks for all the tests! A few minor remarks.

**Comment by mvo5** [comment]
The code that checks for leftover processes reported this on 20.04 after the unit tests ran:
```
2020-11-19T10:19:44.2309454Z tests.invariant: more than one dbus-daemon running
2020-11-19T10:19:44.2310529Z pid:69286 cmdline:/usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only 
```
is one of the tests leaking dbus daemons? or is that a false alarm? (I don't think I saw this particular error in other failed runs though)

**Comment by jhenstridge** [comment]
@mvo5: since that last test run #9497 landed, which included some updates to `tests/lib/tools/cleanup-state` that would likely have fixed the stray dbus-daemon process.  It doesn't seem to have occurred with the new run.

File: tests/main/dbus-activation-session-legacy/task.yaml
Comment (jhenstridge) [reply tests/main/dbus-activation-session-legacy/task.yaml:15]: This was more an attempt at doing a feature test rather than enumerating which systems it would run on.  The idea being to not have to make decisions about future systems added to the test suite.

File: wrappers/dbus.go
Comment (mvo5) [inline wrappers/dbus.go:157]: (nitpick^2) this could be a single `if _, _, err := ...; err != nil {` line

File: wrappers/services.go
Comment (mvo5) [inline wrappers/services.go:236]: (nitpick) maybe mention "dbus" here in the comement now too?

**Review by mvo5 — APPROVED** [review]
Thanks James, this looks great! I have some comments/suggestions/nitpicks inline but feel free to ignore any of them, nothing in there looks serious :)

File: tests/main/dbus-activation-session-legacy/task.yaml
Comment (jhenstridge) [reply tests/main/dbus-activation-session-legacy/task.yaml:48]: I've changed this to match "hello world".  Changing the string would require rebuilding and publishing the `test-snapd-dbus-service` snap, which I'd prefer to avoid in the short term given it is likely to get stuck requiring manual review again.

File: tests/main/dbus-activation-system/task.yaml
Comment (jhenstridge) [reply tests/main/dbus-activation-system/task.yaml:1]: So, the main things that differ between these two are:

1. different location for service activation files
2. session test needs to set up user session for the test user
3. session test needs to execute a bunch of commands with `test.session` to run in the context of that user session.
4. the actual commands are different: passing `--user` to systemctl, calling the session vs. system client snap commands, etc.
5. different plugs/slots to connect.

There'd be more than half the test parameterised, so I don't think it is worth it.

File: tests/main/dbus-activation-session-legacy/task.yaml
Comment (jhenstridge) [reply tests/main/dbus-activation-session-legacy/task.yaml:26]: CentOS 7 is the other major distro in our battery of tests that requires this.

People could also trigger this case if they start a private session bus separate from the systemd managed one.  This could be to isolate some tests, or because they're using some desktop environment that hasn't updated its startup scripts to take into account the new way of doing things.

It seemed easy enough to run the test on all the systems that supported it, rather than just the ancient ones.

**Pull request merged**
